design/aws-ecs-priv-registry-support.md

```diff
@@ -0,0 +1,146 @@
+
+To address issue [#1698](https://github.com/awslabs/aws-cdk/issues/1698), the ECS construct library should provide a way for customers to specify [`repositoryCredentials`](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDefinition.html#ECS-Type-ContainerDefinition-repositoryCredentials) on their container. 
+
+Minimally, this would mean adding a new string field on `ContainerDefinition`, however this doesn't provide any added value in terms of logical grouping or resource creation. We can instead modify the existing ECS CDK construct [`ContainerImage`](https://github.com/awslabs/aws-cdk/blob/master/packages/%40aws-cdk/aws-ecs/lib/container-image.ts) so that repository credentials are specified along with the image they're meant to access.
+
+## General approach
+
+The [`ecs.ContainerImage`](https://github.com/awslabs/aws-cdk/blob/master/packages/%40aws-cdk/aws-ecs/lib/container-image.ts) class already includes constructs for 3 types of images:
+
+* DockerHubImage
+* EcrImage
+* AssetImage
+
+DockerHub images are assumed public, however DockerHub also provides private repositories and there's currently no way to specify credentials for them in the CDK.
+
+There's also no way to specify images hosted outside of DockerHub, AWS, or your local machine. Customers hosting their own registries or using another registry host, like Quay.io or JFrog Artifactory, would need to be able to specify both the image URI and the registry credentials in order to pull their images down for ECS tasks.
+
+Therefore, to support the use of private registries we would need to add credentials to the `DockerHubImage` type (as optional), and add a new `PrivateImage` type which could house credentials (required) and the private registry URI.
+
+We should also have a unified way to define repository credentials which can be passed to the constructors of both image types.  
+
+## Code changes
+
+Given the above, we should make the following additions to support private registry authentication:
+
+1. Define `RepositoryCredentials` interface & class, add to `IContainerImage` 
+2. Update `DockerHubImage` construct to optionally accept and set `RepositoryCreds` 
+3. Add new `PrivateImage` class which implements `IContainerImage` and requires `RepositoryCreds`
+
+
+# Part 1: How to define registry credentials 
+
+For extensibility, we can define a new `IRepositoryCreds` interface with a "credentialsParamater" string and a new `RepositoryCreds` class which satisfies it using specific constructs (e.g., "fromSecret").
+
+```ts
+export interface IRepositoryCreds {
+    readonly credentialsParameter: string
+}
+
+export class RepositoryCreds {
+    public readonly credentialsParameter: string;
+    private readonly secret: secretsManager.Secret;
+
+    public static fromSecret(secret: secretsManager.Secret) {
+      // sets credentialsParameter from secret.secretArn 
+    }
+
+    public static withCredentialsParameter(param: string) {
```

Comment:
Are these literal credentials, or the name of a secret?

If it's the latter, there's no need to make this distinction here. There is a `secretsmanager.SecretString` which can be constructed from a `Secret` object (a secret that gets deployed in the same CDK application) or constructed by hand from a secret that's been prepopulated into the AWS account. In that case, I think the whole `RepositoryCreds` objects can be replaced with a `SecretString` object.

With permissions and everything I agree that the existing API surface there is clunky, but I'm all in favor of simplifying, making the existing objects richer if we can, so that we need fewer classes. I don't want to go into too much design work here in case I'm completely misunderstanding the problem, so I'll shut up now, but get back to me if you want to discuss more.

Reply:
The thought here was this is an escape hatch if someone has the ARN they want to set but don't want to construct a separate `Secret` construct (or maybe in the future if it supports diff kinds of cred resources, this would still work w/o needing to be updated to accommodate a resource type). 

Another diff is that `fromSecret` would add "read" access to the `Secret` to the execution role, while this one would not. But maybe adding it now is jumping the gun?

Re: using `SecretString` : I steered away from it based on the package [REAMDE](https://github.com/awslabs/aws-cdk/tree/master/packages/%40aws-cdk/aws-secretsmanager) warning that using it may lead to the creds surfacing in plain text, but maybe I misunderstood when exactly that could happen.

Reply:
This is an unfortunate case of the docs being too informative and thereby confusing.

They are talking about 2 different `SecretStrings`, btw. The docs are talking about the 
 `SecretString` attribute to create a `Secret`, but the `SecretString` I'm talking about is for *refencing* an *existing* (or created) secret.

What the docs are trying to tell you is that creating a `Secret` via CloudFormation only allows it to be populated by a randomly generated password, because doing otherwise would mean putting the literal secret value you want to put into into the Secret somewhere in your CloudFormation template... defeating the purpose of it being a secret. A randomly generated secret can still be useful for setting the admin login of an RDS instance, where you need *some* secret but you don't really care what the value is.

For our purposes, people are going to have to manually prepopulate a secret into their account, which we can then *read* using the `SecretString` class. It's not documented in that README, which is a shame, but it's documented here: https://integ-docs-aws.amazon.com/CDK/latest/userguide/how_to_get_ext_values.html#passing_secrets_manager

Reply:
> What the docs are trying to tell you is that creating a `Secret` via CloudFormation only allows it to be populated by a randomly generated password

But you can also import an existing secret ([example](https://github.com/awslabs/aws-cdk/tree/master/packages/%40aws-cdk/aws-secretsmanager) in README) into a `Secret` construct, which is the use case I was considering most common. 

> For our purposes, people are going to have to manually prepopulate a secret into their account

Do you mean into AWS Secrets Manager, or the 3rd party account that houses the registry? If the former, what makes `SecretString` more appropriate here?

To clarify: ECS is expecting the actual [ARN of the credential secret](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_RepositoryCredentials.html) here, not the secret value itself. (Apologies if a previous response was misleading.)

Reply:
I was expecting that.

I mean users populate their own Secret, then import into CDK using a name or ARN.

How the credentials are fed into the actual registry is out of scope I think :) (unless it's an ECR registry, in which case it *could* be an autogenerated Secret)

Reply:
Per offline discussion: agreed that a `Secret` construct is appropriate here.